astropy/units/format/utils.py

```diff
@@ -109,6 +109,8 @@ def decompose_to_known_units(unit, func):
                 return decompose_to_known_units(unit._represents, func)
             raise
         return unit
+    else:
+        raise TypeError(f'Incorrect type {type(unit)}')
```

Comment:
The code looks good, but I think the error message itself could be more helpful: since "incorrect type" is just repeating the name of the error. How about "unit argument must be a 'NamedUnit' or 'CompositeUnit', not {type(unit)}."

This is a bit more like for `float`:
```
In [4]: float(None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-b14d5e084cce> in <module>
----> 1 float(None)

TypeError: float() argument must be a string or a number, not 'NoneType'
```